lib/Runtime/Library/WebAssemblyInstance.cpp

```diff
@@ -83,60 +114,57 @@ WebAssemblyInstance::CreateInstance(WebAssemblyModule * module, Var importObject
     }
 
     ScriptContext * scriptContext = module->GetScriptContext();
-    Var* moduleEnvironmentPtr = RecyclerNewArrayZ(scriptContext->GetRecycler(), Var, module->GetModuleEnvironmentSize());
-    Var* memory = moduleEnvironmentPtr + WebAssemblyModule::GetMemoryOffset();
-
+    WebAssemblyEnvironment environment(module);
     WebAssemblyInstance * newInstance = RecyclerNewZ(scriptContext->GetRecycler(), WebAssemblyInstance, module, scriptContext->GetLibrary()->GetWebAssemblyInstanceType());
-    WebAssemblyTable** table = (WebAssemblyTable**)(moduleEnvironmentPtr + module->GetTableEnvironmentOffset());
-    Var* localModuleFunctions = moduleEnvironmentPtr + module->GetFuncOffset();
-    Var* importFunctions = moduleEnvironmentPtr + module->GetImportFuncOffset();
-    LoadGlobals(module, scriptContext, moduleEnvironmentPtr, importObject);
-    LoadImports(module, scriptContext, importFunctions, localModuleFunctions, importObject, memory, table);
-    LoadFunctions(module, scriptContext, moduleEnvironmentPtr, localModuleFunctions);
-    LoadDataSegs(module, memory, scriptContext);
-
-    Js::Var exportsNamespace = JavascriptOperators::NewJavascriptObjectNoArg(scriptContext);
-    LoadIndirectFunctionTable(module, scriptContext, table, localModuleFunctions, importFunctions);
-
-    BuildObject(module, scriptContext, exportsNamespace, memory, table, newInstance, localModuleFunctions, importFunctions);
-
-    uint32 startFuncIdx = module->GetStartFunction();
-    if (startFuncIdx != Js::Constants::UninitializedValue)
+    try
+    {
+        LoadImports(module, scriptContext, importObject, &environment);
+        LoadGlobals(module, scriptContext, &environment);
+        LoadFunctions(module, scriptContext, &environment);
+        LoadDataSegs(module, scriptContext, &environment);
+        LoadIndirectFunctionTable(module, scriptContext, &environment);
+        Js::Var exportsNamespace = BuildObject(module, scriptContext, &environment);
+        JavascriptOperators::OP_SetProperty(newInstance, PropertyIds::exports, exportsNamespace, scriptContext);
+
+        uint32 startFuncIdx = module->GetStartFunction();
+        if (startFuncIdx != Js::Constants::UninitializedValue)
+        {
+            AsmJsScriptFunction* start = environment.GetWasmFunction(startFuncIdx);
+            Js::CallInfo info(Js::CallFlags_New, 1);
+            Js::Arguments startArg(info, (Var*)&start);
+            Js::JavascriptFunction::CallFunction<true>(start, start->GetEntryPoint(), startArg);
+        }
+    }
+    catch (Wasm::WasmCompilationException e)
```

Comment:
The problem is, we use this exception in some of the api that we use for instantiation.
Thus a `WasmCompilationException` can be considered an Instantiation error.